File: ruby/lib/bool/evaluator.rb
module Bool
  class Evaluator
    if RUBY_PLATFORM =~ /java/
      require 'bool_ext'

      def self.new
        Java::Bool::Evaluator.new
      end

    else
      def visit_var(node, vars)
        !!vars.index(node.token.value)
      end

      def visit_and(node, vars)
        evaluate(node.left, vars) && evaluate(node.right, vars)
      end

      def visit_or(node, vars)
        evaluate(node.left, vars) || evaluate(node.right, vars)
      end

      def visit_not(node, vars)
        !evaluate(node.operand, vars)
      end

      def evaluate(node, vars)
        node.accept(self, vars)
      end
    end
  end
end

File: javascript/lib/evaluator.js
module.exports = function Evaluator() {
  var self = this;

  this.visit_var = function(node, vars) {
    return vars.indexOf(node.token.value) != -1;
  };

  this.visit_and = function(node, vars) {
    return self.evaluate(node.left, vars) && self.evaluate(node.right, vars);
  };

  this.visit_or = function(node, vars) {
    return self.evaluate(node.left, vars) || self.evaluate(node.right, vars);
  };

  this.visit_not = function(node, vars) {
    return !self.evaluate(node.operand, vars);
  };

  this.evaluate = function(node, vars) {
    return node.accept(self, vars)
  }
};
